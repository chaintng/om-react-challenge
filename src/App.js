import React, { Component } from 'react';
import { connect } from 'react-redux';
import fetch from 'isomorphic-fetch';

import { summaryDonations } from './helpers';
import Card from './components/Card';

export default connect((state) => state)(
  class App extends Component {
    constructor(props) {
      super(props);

      this.state = {
        charities: [],
        selectedAmount: 10,
      };

      this.handlePay = this.handlePay.bind(this);
    }

    componentDidMount() {
      const self = this;
      fetch('http://localhost:3001/charities')
        .then(function(resp) { return resp.json() })
        .then(function(data) {
          self.setState({ charities: data }); });

      fetch('http://localhost:3001/payments')
        .then(function(resp) { return resp.json() })
        .then(function(data) {
          self.props.dispatch({
            type: 'UPDATE_TOTAL_DONATE',
            amount: summaryDonations(data.map((item) => (item.amount))),
          });
        });
    }

    handlePay(id, amount, currency) {
      const self = this;
      fetch('http://localhost:3001/payments', {
        method: 'POST',
        body: `{ "charitiesId": ${id}, "amount": ${amount}, "currency": "${currency}" }`,
      })
        .then(function (resp) {
          return resp.json();
        })
        .then(function () {
          self.props.dispatch({
            type: 'UPDATE_TOTAL_DONATE',
            amount,
          });
          self.props.dispatch({
            type: 'UPDATE_MESSAGE',
            message: `Thanks for donate ${amount}!`,
          });

          setTimeout(function () {
            self.props.dispatch({
              type: 'UPDATE_MESSAGE',
              message: '',
            });
          }, 2000);
        });
    }

    render() {
      const { donate, message } = this.props;
      const cards = this.state.charities.map((item, i) => (
        <Card key={i} {...item} handlePay={this.handlePay}/>
      ));

      const style = {
        color: 'red',
        margin: '1em 0',
        fontWeight: 'bold',
        fontSize: '16px',
        textAlign: 'center',
      };

      return (
        <div>
          <h1>Tamboon React</h1>
          <p>All donations: {donate}</p>
          <p style={style}>{message}</p>
          {cards}
        </div>
      );
    }
  }
);
