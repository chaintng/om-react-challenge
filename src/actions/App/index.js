import fetch from 'isomorphic-fetch';
import config from '~/config';
import {summaryDonations} from '~/helpers';
import Promise from 'bluebird';

export const initializeCharities = (charities) => {
  return {
    type: 'INITIALIZE_CHARITIES',
    charities,
  };
};

export const updateTotalDonate = (amount) => {
  return {
    type: 'UPDATE_TOTAL_DONATE',
    amount,
  };
};

export const updateMessage = (visible, level, message) => {
  return {
    type: 'UPDATE_MESSAGE',
    notification: {
      visible,
      level,
      message,
    },
  };
};

export const hideMessage = (visible) => {
  return {
    type: 'HIDE_MESSAGE',
    notification: {
      visible,
    },
  };
};

export const searchKeywordChange = (searchKeyword) => {
  return {
    type: 'SEARCH_KEYWORD_CHANGE',
    searchKeyword,
  };
};

// redux thunk

export const payDonation = ({charitiesId, amount, currency}) => {
  return async (dispatch) => {
    try {
      const response = await fetch(`${config.BACKEND_ENDPOINT}/payments`, {
        method: 'POST',
        body: JSON.stringify({ charitiesId, amount, currency }),
        headers: {
          'Content-Type': 'application/json'
        }
      });

      if (response.status === 404) {
        throw response;
      }

      dispatch(updateTotalDonate(amount));
      dispatch(updateMessage(true, 'success', `Thanks for donate ${amount}!`));

    } catch (e) {
      console.error(e);
      dispatch(updateMessage(true, 'error', 'Sorry, there is some error during payment. Please try again later.'));
    } finally {
      setTimeout(function () {
        dispatch(hideMessage(false));
      }, 2000);
    }
  };
};

export const hydrateAppData = () => {
  return async (dispatch) => {
    const [charities, payments] = await Promise.all([
      fetch(`${config.BACKEND_ENDPOINT}/charities`).then((resp) => resp.json()),
      fetch(`${config.BACKEND_ENDPOINT}/payments`).then((resp) => resp.json()),
    ]);

    dispatch(initializeCharities(charities));
    dispatch(updateTotalDonate(summaryDonations(payments.map((item) => (item.amount)))));
  };
};