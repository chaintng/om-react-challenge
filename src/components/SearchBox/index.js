import React, {Component} from 'react';
import styled from 'styled-components';

const Wrapper = styled.div`
  min-height: 100px;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 1.4em;
`;

class SearchBox extends Component {
  render() {
    return (
      <Wrapper>
        <div style={{marginRight: '10px'}}>Search Charities:</div><input type="text" />
      </Wrapper>
    );
  }
}

export default SearchBox;