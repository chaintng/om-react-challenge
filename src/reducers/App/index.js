const defaultState = {
  charities: [],
  donate: 0,
  message: '',
};

function app(state = defaultState, action) {
  switch (action.type) {
    case 'INITIALIZE_CHARITIES':
      return { ...state,
        charities: action.charities,
      };
    case 'UPDATE_TOTAL_DONATE':
      return { ...state,
        donate: state.donate + action.amount,
      };
    case 'UPDATE_MESSAGE':
      return { ...state,
        message: action.message,
      };

    default:
      return state;
  }
}

export default app;