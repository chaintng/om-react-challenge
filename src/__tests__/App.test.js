import React from 'react';
import { shallow, mount } from 'enzyme';
import thunk from 'redux-thunk';
import 'isomorphic-fetch';
import fetchMock from 'fetch-mock';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import App from '~/App';
import SearchBox from '~/components/SearchBox';


describe('<App />', () => {
  test.only('when SearchKeywordChange, it dispatch the right action', () => {
    const mockStore = configureStore([thunk]);
    const initialState = {
      charities: [],
      donate: 0,
      message: '',
      notification: null,
    };
    const store = mockStore(initialState);

    fetchMock.get('*', {hello: 'world'});

    const wrapper = shallow(
      <App store={store} />
    );

    const component = wrapper.dive();

    component.find(SearchBox).dive().find('input').simulate('change', { target: { value: 'kru' }});

    expect(store.getActions()).toMatchSnapshot();
  });
});
