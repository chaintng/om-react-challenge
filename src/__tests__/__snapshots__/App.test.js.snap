// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> when SearchKeywordChange, it dispatch the right action 1`] = `
Array [
  Object {
    "searchKeyword": "kru",
    "type": "SEARCH_KEYWORD_CHANGE",
  },
]
`;
